internal/scheduler/nodedb.go

```diff
@@ -0,0 +1,332 @@
+	NodeTypes map[string]*schedulerobjects.NodeType
+	// Resources allocated by the scheduler to in-flight jobs,
+	// i.e., jobs for which resource usage is not yet reported by the executor.
+	AssignedByNode map[string]schedulerobjects.AssignedByPriorityAndResourceType
+	// Map from job id to the set of nodes on which that job has been assigned resources.
+	// Used to clear AssignedByNode once jobs start running.
+	NodesByJob map[uuid.UUID]map[string]interface{}
+	// Map from node id to the set of jobs that have resourced assigned to them on that node.
+	// Used to clear AssignedByNode once jobs start running.
+	JobsByNode map[string]map[uuid.UUID]interface{}
+	// Mutex to control access to AssignedByNode, NodesByJob, and JobsByNode.
+	mu sync.Mutex
+}
+
+func (nodeDb *NodeDb) SelectAndBindNodeToPod(jobId uuid.UUID, req *schedulerobjects.PodRequirements) (*PodSchedulingReport, error) {
+	report, err := nodeDb.SelectNodeForPod(jobId, req)
+	if err != nil {
+		return nil, err
+	}
+
+	if report.Node != nil {
+		err = nodeDb.BindNodeToPod(jobId, req, report.Node)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return report, nil
+}
+
+// SelectAndBindNodeToPod selects a node on which the pod can be scheduled,
+// and updates the internal state of the db to indicate that this pod is bound to that node.
+func (nodeDb *NodeDb) SelectNodeForPod(jobId uuid.UUID, req *schedulerobjects.PodRequirements) (*PodSchedulingReport, error) {
+	// Collect all node types that could potentially schedule the pod.
+	nodeTypes, numExcludedNodeTypesByReason, err := nodeDb.NodeTypesMatchingPod(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store number of nodes excluded by reason.
+	numExcludedNodesByReason := make(map[string]int)
+
+	// The dominant resource is the one for which the pod requests
+	// the largest fraction of available resources.
+	// For efficiency, the scheduler only considers nodes with enough of the dominant resource.
+	dominantResourceType := nodeDb.dominantResource(req)
+
+	// Create a report to be returned to the caller.
+	report := &PodSchedulingReport{
+		Timestamp:                    time.Now(),
+		JobId:                        jobId,
+		Req:                          req,
+		DominantResourceType:         dominantResourceType,
+		NumMatchedNodeTypes:          len(nodeTypes),
+		NumExcludedNodeTypesByReason: numExcludedNodeTypesByReason,
+		NumExcludedNodesByReason:     numExcludedNodesByReason,
+	}
+
+	// Iterate over candidate nodes.
+	txn := nodeDb.Db.Txn(false)
+	it, err := NewNodeTypesResourceIterator(
+		txn,
+		dominantResourceType,
+		req.Priority,
+		nodeTypes,
+		req.ResourceRequirements.Requests[v1.ResourceName(dominantResourceType)],
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	for obj := it.Next(); obj != nil; obj = it.Next() {
+		node := obj.(*schedulerobjects.Node)
+		if node == nil {
+			break
+		}
+		// TODO: Use the score when selecting a node.
+		nodeDb.mu.Lock()
```

Comment:
what's the mutex protecting here? Let's add a comment.

Reply:
Agreed